### PR TITLE
valid CLDF with working conversion to sqlite

### PR DESCRIPTION
**Pull request by xrotwang**
Working with the ~2Mio rows in `phones.csv` via `pycldf` might be kind of cumbersome due to the slowness. Now you could pay the price once
```shell
$ time cldf createdb cldf/Generic-metadata.json doreco.sqlite
INFO    <cldf:v1.0:Generic at cldf> loaded in doreco.sqlite

real	3m25,673s
user	3m22,994s
sys	0m2,521s
```
and get fast access later
```shell
$ time sqlite3 doreco.sqlite 'select count(*) from `phones.csv`'
1866499

real	0m0,021s
user	0m0,009s
sys	0m0,013s
```

### CODE REVIEW

File: cldfbench_doreco.py [hunk 1]
```diff
@@ -171,15 +172,15 @@ def create_schema(self, cldf):
             'LanguageTable',
             {
                 'name': 'Source',
-                'datatype': 'str',
+                'datatype': 'string',
```
Comment (xrotwang) [inline cldfbench_doreco.py:415]: @Tarotis the CSVW datatype is called [string](https://csvw.readthedocs.io/en/latest/datatypes.html#csvw.datatypes.string).

File: cldfbench_doreco.py [hunk 2]
```diff
@@ -314,172 +315,174 @@ def create_schema(self, cldf):
...
             },
             {
                 'name': 'spk_code',
-                'datatype': 'int',
+                'datatype': 'string',
             },
             {
                 'name': 'spk_age',
-                'datatype': 'int',
+                'datatype': 'string',
```
Comment (xrotwang) [inline cldfbench_doreco.py:637]: We might want to normalize the content of this column to actually get integers here.

Comment (FredericBlum) [reply cldfbench_doreco.py:637]: Currently, there are some lists of integers, separated by slahes. Is there some way to adapt to this structure? `spk_code` and some other columns are structured identically.

**Comment by FredericBlum** [comment]
The SQL database seems great for all cases that use the data directly. But for all kinds of preprocessing, I would still have to loop through the individual rows, right?

File: cldfbench_doreco.py (hunk 2 above)
Comment (xrotwang) [reply cldfbench_doreco.py:637]: We could try
```json
{
  "separator": "/",
  "datatype": "int"
}
```
and then assign `row['spk_age'].split('/')`.

**Comment by xrotwang** [comment]
What kinds of preprocessing do you mean?

**Comment by FredericBlum** [comment]
For my study, for example, I calculate the log-count of phonemes per second in all interpausal units, or annotate the position of phonemes within their word. This is why I assume that looping through the data-frame is necessary anyway.

**Comment by FredericBlum** [comment]
All calculations involve a lot of looking at preceding and succeding data points. Computationally heavy, but conceptually simple I guess.

**Comment by xrotwang** [comment]
If narrowing down to "interpausal units" is simple to do with a SQL query, you might be able to cut down the amount of stuff to loop through considerably. From my experience, trial-and-error coding is a lot less effective if iterations take 10s of seconds.

**Comment by FredericBlum** [comment]
I wouldn't know, because I have no experience with SQL. Two examples:
Interpausal Units: Add up all durations of elements between two rows where `ph == <p:>`, divide the number of phones through this value
Position: Does the preceding row have the same wd_ID? If no, the current phone is word-initial, if yes, it is not.

**Comment by xrotwang** [comment]
Ah, "adding up all the durations" reminds me that we really should model these using proper duration datatypes!
And I admit that my first go at this would be in Python, too :)

**Comment by xrotwang** [comment]
Ah, `duration` seems to be just a `decimal` value in milliseconds - correct?

**Comment by FredericBlum** [comment]
Yes, exactly. Usually the first thing I do with this table is to convert everything in milliseconds.

**Review by FredericBlum — APPROVED** [review]

**Pull request merged**